lib/retry.js

```diff
@@ -4,37 +4,67 @@ const sleep = msec => new Promise(resolve => {
   setTimeout(resolve, msec);
 });
 
-const promisify = fn => (...args) => (
-  new Promise((resolve, reject) => (
-    fn(...args, (err, data) => (
-      err ? reject(err) : resolve(data)
-    ))
-  ))
-);
+const promisify = (fn, args) => new Promise((resolve, reject) => {
+  if (!args) args = [];
+  args.push((err, data) => (err ? reject(err) : resolve(data)));
+  fn(...args);
+});
 
-const retry = async (times, fn, ...args) => {
-  let tries = 0;
-  let done = false;
-  let res = null;
-  const fnR = fn;
-  if (args[args.length - 1] instanceof Function) {
-    fnR = promisify(fn);
-    args.pop();
+const deepClone = obj => {
+  if (obj === null || typeof obj !== 'object') {
+    return obj;
+  } else if (Array.isArray(obj)) {
+    const result = [];
+    for (const data of obj) {
+      result.push(deepClone(data));
+    }
+    return result;
+  } else {
+    const result = {};
+    const keys = Object.keys(obj);
+    for (const key of keys) {
+      result[key] = deepClone(obj[key]);
+    }
+    return result;
   }
-  while (!done && tries < times) {
-    tries++;
+};
+
+const expectInterval = async data => {
+  const interval = data.exponential ?
+    data.interval * data.multiplier :
+    data.interval;
+  if (interval) await sleep(interval);
+};
+
+const retry = async (fn, args = [], config = {}) => {
+  const retries = config.retries || 3;
+  const interval = config.interval || 0;
+  const exp = Object.prototype.hasOwnProperty.call(config, 'exponential') ?
+    config.exponential :
+    true;
+  const multiplier = config.multiplier || 2;
+  const presentCb = typeof config.onAttemptFail === 'function';
+  const onAttemptFail = presentCb ? config.onAttemptFail : expectInterval;
+
+  for (let i = 0; i < retries; i++) {
     try {
-      res = await fnR(...args);
-      done = true;
+      const result = !config.isCb ?
+        await fn(...args) :
+        await promisify(fn, deepClone(args));
+      return result;
     } catch (err) {
-      if (tries >= times) {
-        throw err;
-      } else {
-        console.dir({ error: err.message, tries });
-      }
+      if (retries === i + 1) throw err;
+      const result = await onAttemptFail({
+        err,
+        attempt: i + 1,
+        retries,
+        interval,
+        exponential: exp,
+        multiplier,
+      });
+      if (presentCb && !result) return;
     }
   }
-  return res;
 };
 
-module.exports = { retry, sleep };
+module.exports = { retry };
```

Comment:
maybe change it for `module.exports = retry;` because we exports only one item?

Reply:
Fixed, thank you.